### PR TITLE
fix(extmarks): problems with folded virtual lines

### LINKED ISSUE #17027: virt_lines overwrite foldtext on folded lines
### Neovim version (nvim -v)

0.6.1

### Vim (not Nvim) behaves the same?

do not apply

### Operating system/version

linux (Archlinux)

### Terminal name/version

kitty 0.24.0

### $TERM environment variable

xterm-kitty

### Installation

pacman

### How to reproduce the issue
```
nvim --clean

:set fdm=indent
ifirst line
        indented line
        another indented line<Esc>
:lua vim.api.nvim_buf_set_extmark(0, vim.api.nvim_create_namespace("vline"), 1, 0, { virt_lines={{{"Virt Line", "Comment"}}}, virt_lines_above=true })
za
```

### Expected behavior

the virtual line is not shown as referring to a folded text.
I can see the `foldtext`.

### Actual behavior

The `foldtext` is hidden by an empty virtual line.

This happens if the virtual line is `above` the first line of the folded text.
When `virt_lines_above=false` and the virtual line is `below` last line of the folded text, the virtual line is shown below the `foldtext` even when the text is folded (not sure is this the intended behaviour).

### LINKED ISSUE #19557: Virtual lines and fold text conflict
### Neovim version (nvim -v)

NVIM v0.8.0-dev-724-gbdbf84303

### Vim (not Nvim) behaves the same?

No, no virtual lines in Vim

### Operating system/version

Arch Linux

### Terminal name/version

Kitty 0.25.2

### $TERM environment variable

xterm-kitty

### Installation

build from source

### How to reproduce the issue

```init.lua
vim.wo.foldenable = true
vim.wo.foldlevel = 1
vim.wo.foldmethod = "marker"
vim.wo.number = true

vim.api.nvim_create_autocmd("BufEnter", {
    group = vim.api.nvim_create_augroup("folds", {}),
    pattern = "<buffer>",
    callback = function()
        vim.api.nvim_buf_set_text(0, 0, 0, 0, 0, {
            "",
            "{{{",
            "hello",
            "}}}",
        })

        local ns = vim.api.nvim_create_namespace "folds"
        vim.api.nvim_buf_set_extmark(0, ns, 1, 0, {
            virt_lines_above = true,
            virt_lines = { { { "virtual text", "Error" } } },
        })
    end,
})
```

Open Neovim and toggle folding with `zm`

### Expected behavior

The virtual text should be displayed first, it is outside the fold.
Then the fold text should be displayed.

### Actual behavior

Neither the virtual text, nor the fold text is displayed, there is just an empty virtual line.
The cursor is now able to go out of bound of the buffer into the end of buffer section.

### LINKED ISSUE #21837: Folded line (and cursor) with a virtual line attached is not drawn (correctly)
### Describe the bug

Folded line with a virtual line attached is not drawn. The cursor can still be moved to it but it's drawn position is incorrect offset by the number of virtual lines:
image

### Steps to reproduce

test.lua:
```lua
local ns = vim.api.nvim_create_namespace("ns")
vim.o.number = true
vim.api.nvim_buf_set_extmark(0, ns, 2, 0, { virt_lines = {{{"virt_line below 3", ""}}} })
vim.api.nvim_buf_set_extmark(0, ns, 3, 0, { virt_lines_above = true, virt_lines = {{{"virt_line above 4", ""}}} })
vim.cmd.norm("4Gzfj")
```

    nvim --clean -S test.lua test.lua

### Expected behavior

Folded line is drawn correctly.

Looking into a fix.

### Neovim version (nvim -v)

NVIM v0.9.0-dev-701+g34b973b1d9

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

linux

### Terminal name/version

st

### $TERM environment variable

st-256color

### Installation

AUR

### PR DESCRIPTION
**Pull request by luukvbaal**
Problem:    When a folded line has virtual lines attached, the following
            problems occur:
              - The virtual lines are drawn empty.
              - The 'foldtext' line is drawn empty.
              - The cursor is drawn incorrectly.
Solution:   Check whether virtual lines belong to a folded line.

Fix #17027
Fix #19557
Fix #21837

### CODE REVIEW

**Review by lewis6991 — APPROVED** [review]
Awesome!

File: src/nvim/plines.c [hunk 1]
```diff
@@ -51,7 +51,8 @@ int plines_win(win_T *wp, linenr_T lnum, bool winheight)
 /// @return Number of filler lines above lnum
 int win_get_fill(win_T *wp, linenr_T lnum)
 {
-  int virt_lines = decor_virt_lines(wp, lnum, NULL);
+  foldinfo_T fi = fold_info(wp, lnum);
```
Comment (zeertzjq) [inline src/nvim/plines.c:54]: `win_get_fill()` is used quite often. Will using `fold_info()` here cause performance problems?

Comment (zeertzjq) [reply src/nvim/plines.c:54]: `hasFoldingWin()` with `cache` is used for diff:
https://github.com/neovim/neovim/blob/0f52e2c849cbd923fc30b8bbef088b6f7642aeab/src/nvim/diff.c#L2137-L2140

Comment (luukvbaal) [reply src/nvim/plines.c:54]: The cache seems to work fine for the added tests, I can't judge right now if it will always be correct.

**Comment by zeertzjq** [comment]
Pushed a fixup commit to avoid calling `hasFoldingWin()` when where are no virt lines

File: test/functional/ui/fold_spec.lua [hunk 2]
```diff
@@ -1852,6 +1853,127 @@ describe("folded lines", function()
...
+            [2:---------------------------------------------]|
+            [2:---------------------------------------------]|
+            [2:---------------------------------------------]|
+            [3:---------------------------------------------]|
+          ## grid 2
+            virt_line above line 1                       |
+            ^line 1                                       |
+            line 2                                       |
+            virt_line below line 2                       |
+            {5:+--  2 lines: line 3}                         |
```
Comment (zeertzjq) [inline test/functional/ui/fold_spec.lua:1958]: This looks wrong.

Comment (zeertzjq) [reply test/functional/ui/fold_spec.lua]: Fixed in another fixup commit

**Review by zeertzjq — APPROVED** [review]

**Pull request merged**